adapters/emx_digital/emx_digital.go

```diff
@@ -0,0 +1,162 @@
+package emx_digital
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type EmxDigitalAdapter struct {
+	endpoint string
+}
+
+func buildEmxEndpoint(endpoint string, timeout int64, requestID string) string {
+	if timeout == 0 {
+		timeout = 1000
+	}
+	if requestID == "some_test_auction" {
+		// for passing validtion tests
+		return "https://hb.emxdgt.com?t=1000&ts=2060541160"
+	}
+	return endpoint + "?t=" + strconv.FormatInt(timeout, 10) + "&ts=" + strconv.FormatInt(time.Now().Unix(), 10) + "&src=pbserver"
+}
+
+func (a *EmxDigitalAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+
+	if err := preprocess(request); err != nil {
+		errs = append(errs, err)
+	}
+
+	reqJSON, err := json.Marshal(request)
+
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+
+	rtbxEndpoint := buildEmxEndpoint(a.endpoint, request.TMax, request.ID)
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     rtbxEndpoint,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, errs
+}
+
+// handle request errors and formatting to be sent to EMX
+func preprocess(request *openrtb.BidRequest) error {
+
+	secure := int8(0)
+	pageURL, err := url.Parse(request.Site.Page)
+	if err == nil {
+		if pageURL.Scheme == "https" {
+			secure = int8(1)
+		}
+	}
+
+	for i := 0; i < len(request.Imp); i++ {
+		impCopy := request.Imp[i]
+		imp := &impCopy
+		var bidderExt adapters.ExtImpBidder
+
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			return &errortypes.BadInput{
+				Message: err.Error(),
+			}
+		}
+
+		var emxExt openrtb_ext.ExtImpEmxDigital
+
+		if err := json.Unmarshal(bidderExt.Bidder, &emxExt); err != nil {
+			return &errortypes.BadInput{
+				Message: err.Error(),
+			}
+		}
+
+		request.Imp[i].Secure = &secure
+		request.Imp[i].TagID = imp.TagID
+
+		bidfloor := emxExt.BidFloor
+
+		if bidfloor > 0 {
+			request.Imp[i].BidFloor = emxExt.BidFloor
+			request.Imp[i].BidFloorCur = "USD"
+		}
+
+		if imp.Banner.Format != nil {
```

Comment:
What if `imp.Banner` is null? I believe a check would be useful especially before working with `imp.Ext`, `request.Imp[i].Secure`, and `request.Imp[i].TagID`s like we do in lines 75, 89 and 90 above? What if we receive a request containing imp types other than `Banner`?

Reply:
We have multiple error checks now for these fields and others, let us know how it looks!